File: Data Structures & Algorithms/length-of-last-word/submission-2.kt
class Solution {
    fun lengthOfLastWord(s: String): Int {
        var ptr = s.length - 1
        while (ptr >= 0 && s[ptr] == ' ') {
            ptr -= 1
        }

        var result = 0
        while (ptr >= 0 && s[ptr] != ' ') {
            result += 1
            ptr -= 1
        }

        return result
    }
}

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-2.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total_sum = sum(nums)
        if total_sum % 2 != 0:
            return False
        half_sum = total_sum / 2

        memo = {}

        def dfs(index: int, reminder: int):
            if reminder == 0:
                return True
            if index >= len(nums) or reminder < 0:
                return False

            key = f"{index}-{reminder}"
            if key in memo:
                return memo[key]

            left = dfs(index + 1, reminder - nums[index])
            right = dfs(index + 1, reminder)

            result = left or right
            memo[key] = result

            return result

        return dfs(0, half_sum)
            